### ISSUE TITLE
Webadmin technology redesign

### ISSUE THREAD

**Issue by sphinxc0re** [cuberite/cuberite#3003]
I believe the webadmin needs a complete redesign, technology-wise. Modern web applications supply static files once and just once. After that, JavaScript kicks in and handles every data request which is then rendered by the client in real time.

The current webadmin still renders most of the client on the server side. We could connect this with implementing a REST API into the server. We could even throw the whole webadmin out of the server and reimplement it into a separate plugin.

**Comment by NiLSPACE** [comment]
I'm not sure. All the webadmin does is provide the proper layout and stuff. Using javascript and REST API calls is up to the plugins.

Making the webadmin as a plugin is possible, but would require all the plugins that are using it to be changed. I don't think it's worth it.

**Comment by madmaxoft** [comment]
I'm currently rewriting most of the HTTP stuff and most of webadmin will get changed with that rewrite.

NiLSPACE <notifications@github.com> wrote:

>I'm not sure. All the webadmin does is provide the proper layout and
>stuff. Using javascript and REST API calls is up to the plugins.
>
>Making the webadmin as a plugin is possible, but would require all the
>plugins that are using it to be changed. I don't think it's worth it.
>
>---
>Reply to this email directly or view it on GitHub:
>https://github.com/cuberite/cuberite/issues/3003#issuecomment-183367696

-- 
Sent from my Android phone with K-9 Mail. Please excuse my brevity.

**Comment by Ajusa** [comment]
I would be MORE than happy to do a JavaScript.
Jokes aside, if there is a rest API available, I would be willing to help with the JS.

**Comment by NiLSPACE** [comment]
I think we first need to determine if it's actually worth it, or if there are problems that could arise.

For example, what would happen if I make a plugin that adds a webtab in the webadmin which uses a little javascript that schedules a function every X seconds. Currently a complete new page gets loaded, so the function is aborted, but when a piece of javascript changes the tab meaning that the page doesn't get reloaded won't the javascript crash?

Naturally the user won't get to see this problem unless the plugin uses something like alert in case something is wrong, but in case 2 plugins use something that can conflict (for example the same class names in the HTML) things can get weird.

**Comment by SphinxC0re** [comment]
In your case you won't even need the page reload because you could just request new data asynchronously via ajax. Also we could enclose the supplied js files with closures:
```javascript
(function() {
// code
})();

**Comment by NiLSPACE** [comment]
But would that fix scheduled functions? Concider the chat page in the Core. It has a function that is called every X seconds that checks for new chat messages. If I switch page that will still be running. In fact, when I go back to the page later again I'd have the callback called twice. How will that be handled?

**Comment by Ajusa** [comment]
I don't see an issue with that. If the callback is called twice, then what
happens? The chat area re renders twice?

I don't think that there are many problems with this. I have some
experience with this kind of stuff and I would like to contribute.

Maybe I am not understanding something...

**Comment by NiLSPACE** [comment]
It's possible that nothing happens (or at least to the user). It's also possible that the message will be displayed twice depending on the code of the Core.

What would be certain though is that it sends a network request to the server which isn't pretty.

**Comment by SphinxC0re** [comment]
Usually you won't ask the server for new data in an interval. Instead you would most probably poll for a new message (or more) and just wait for a response until a new message arrives on the server

**Comment by Ajusa** [comment]
I think a good idea is to make an API that anybody can use. That way we could have alternative web interfaces for other developers to make. There could be one official one, and then several community 'plugin' ones. 

I don't think it is like that right now?
I could be wrong

**Comment by NiLSPACE** [comment]
Is it perhaps possible to sandbox javascript code? Kinda like how you can set the environment in Lua using setfenv?

**Comment by madmaxoft** [comment]
@NiLSPACE I still don't understand what you're afraid of. It's not like each plugin would load its JS code into the same global space, it would still be requested on a per-page basis.

**Comment by NiLSPACE** [comment]
If an admin loads the chat page it means it will check for new messages every X seconds righ? If I then change page won't it still be checking for new messages?

**Comment by madmaxoft** [comment]
I believe we should still keep some form of pagination. Think GitHub - it has pages for individual areas of work, but each such page can do many things using JS (such as displaying and adding comments, editing labels etc. for an issue, but the repo viewer is a completely separate page)

**Comment by LogicParrot** [comment]
>If I then change page won't it still be checking for new messages?

Unsanboxed Javascript could fool you into thinking you've switched a page when in fact you haven't, via html5 pushstate. Just saying.

**Comment by worktycho** [comment]
JavaScript has no concept of sandboxing other than iframes. The best we can do is separate global scopes. Plugins either need to be cooperative, or we need to compile something else to Js if we want sandboxing.

**Comment by SphinxC0re** [comment]
Coffeescript?

**Comment by LogicParrot** [comment]
In my humble opinion, doing this securely is not easy, and considering the current state of Cuberite, it would be better to channel the manpower to fixing problems in existing features before introducing even more features. This sounds interesting as a future project though.

**Comment by LogicParrot** [comment]
Is security actually needed? Wouldn't a bad plugin cause server chaos anyways?

**Comment by LogicParrot** [comment]
- Is security actually needed? Wouldn't a bad plugin cause server chaos anyways?
 - Is this worth it?

**Comment by Ajusa** [comment]
I think it is worth it. There are many JavaScript developers, so this could get implemented quickly if there is an API. I really want it to be modular, SK there can be multiple dashboards, so that peopl e can change what they want.

**Comment by Schwertspize** [comment]
Additionally it would be great eg to be able to have some kind of dashboard where plugins can "register" their stuff eg get a tile on it to fill with whatever the plugin is doing.... Like core could display online players, chat,... Where performancePages could add a graph about the performance....

**Comment by yangm97** [comment]
For the back end, I think the webadmin should be its own plugin, just like core. Cuberite would then only need to handle web-serving (binding a port for each plugin? Single port for everything but with fake directories, like cuberiteserver.com/pluginname/? IDK). Please note this wouldn't replace the webadmin API for plugins, this would just allow plugins like dynmap to show public facing pages

The front end side of things should be quite simple to do. I think using the twitter bootstrap should allow for better accessibility. Current UI is not responsive at all (i.e. doesn't know how to handle a phone screen). Tables? Who made this? A dinosaur? 😨
And finally, it should make it easier to code the webadmin itself and it's plugins, since bootstrap handles most of the heavy lifting, and is easy to use even if you're not a designer.

Any thoughts? I could rebuild the frontend myself, even though I personally use Zurb Foundation, I admit Bootstrap is more popular and has more building blocks, thus making it easier to use than Foundation.

**Comment by Cl1608Ho** [comment]
Uhm because I read the thread about the chat.... First of all you can unregister any setInterval() you ever used and it is advisable to do so when you aren't in this page anymore. Secondly, because sphinxc0re mentioned long-polling, I actually think WebSockets are the solution, so as long as you browse the page, you have a WebSocket connection with the server and every plugin can pass data here, which is stuff like a new player joined or a there is a chat message => the web page should display it. 

A high level implementation of this is socket.io (when we want to use it, every plugin would get its own namespace) but we can do the same with plain WebSockets I think.

**Comment by madmaxoft** [comment]
I may be a dinosaur, but the webadmin, as it is, will render on a five year old tablet, as well as in text-mode browsers, and it will render in an eye-blink. Which couldn't be said about all those flashy new javascript thingies. If we really wanted a responsive UI, we could listen for the screen size http header server-side and adjust the css.

The problem with JS and the new frameworks is that they are new and fast-changing. If we decide for a cutting edge technology today, half the browsers won't support it, in a year most (new) browsers will, but will be painfully slow, another year for a reasonable performance, and by then the framework will be so painfully obsolete that we'll be looking for a new one. Tables have been in HTML since the beginning and they will still be there in 10 years.

**Comment by yangm97** [comment]
> but the webadmin, as it is, will render on a five year old tablet

This wouldn't change, in fact it would have a proper interface for any tablet, phone or desktop without messing with JS, there are CSS queries for that, graceful degradation and progressive enhancement techniques etc. JS is not needed for design, I will make sure these continue to render gracefully even with javascript disabled. 

Using bootstrap CSS doesn't mean using bootstrap JS. And divs have been around for quite a long too, and work just as well as tables. Server side detection of device screen is unreliable and should be avoided.

Webadmin is a quite simple page and I don't want to go over the top with it. I could rebuild it using pure CSS, but I believe providing developers with a well known web framework for building the interface of their plugins, while also using AJAX, websockets or something to render pages without reloading the whole website is going to give plugin developers and users a better experience than the current one, and it's not hard to maintain fallbacks (like iframes for the AJAX/websockets thing). 
5 years from now we have iOS devices running iOS9, which is not the latest but capable still, and Android 4.0 running (the latest I assume) chrome version.

**Comment by Pokechu22** [comment]
Tables should not be used for non-tabular content, _period_.  Using tables for non-tabular content is a hack at best.

**Comment by madmaxoft** [comment]
[rant] Au contraire, tables are the only way to lay out the page that simply works across *all* browsers, *period*. [/rant]

**Comment by yangm97** [comment]
It's not working on safari, neither does on phones.

**Comment by madmaxoft** [comment]
The "not working" part is the CSS, not the table itself.

**Comment by madmaxoft** [comment]
But hey, I'm a dinosaur anyway, so whatever you guys want. If you can spec out a nice interface for everything, I'll be more than happy to provide the Lua bindings needed.

**Comment by sphinxc0re** [comment]
Wow...! That had an influence on xofts mind 😲 everyone calm down, please 😄

**Comment by yangm97** [comment]
@madmaxoft sorry for the dinosaur thing.

**Comment by Pokechu22** [comment]
> [rant] Au contraire, tables are the only way to lay out the page that simply works across all browsers, period. [/rant]

Quoting the [HTML5 standard, section 4.9: Tabular Data](http://www.w3.org/TR/html5/tabular-data.html):

> Tables should not be used as layout aids. Historically, many Web authors have tables in HTML as a way to control their page layout making it difficult to extract tabular data from such documents. In particular, users of accessibility tools, like screen readers, are likely to find it very difficult to navigate pages with tables used for layout. If a table is to be used for layout it must be marked with the attribute role="presentation" for a user agent to properly represent the table to an assistive technology and to properly convey the intent of the author to tools that wish to extract tabular data from the document.
>
> > **Note**: There are a variety of alternatives to using HTML tables for layout, primarily using CSS positioning and the CSS table model. [\[CSS\]](http://www.w3.org/TR/html5/references.html#refsCSS)

And the [HTML4.1 spec](http://www.w3.org/TR/html401/struct/tables.html):

> Tables should not be used purely as a means to layout document content as this may present problems when rendering to non-visual media. Additionally, when used with graphics, these tables may force users to scroll horizontally to view a table designed on a system with a larger display. To minimize these problems, authors should use [style sheets](http://www.w3.org/TR/html401/present/styles.html) to control layout rather than tables.

That says a bunch of things, but the key part I'll note is that tables _don't_ work on all types of browsers and can actually make organization worse on some (mainly text-to-speech).

That doesn't mean you can't use tables, but it's probably not the best technique.

**Comment by yangm97** [comment]
I’m thinking about making a visual makeup to the webadmin – mostly due the viewport issues some mobile devices have been having. Nothing great, I plan to mimick most of cuberite/cuberite.github.io.

Anyone against this?

**Comment by mathiascode** [comment]
Going to close this issue, since the Webadmin recently received improvements to make it adapt better to smaller screens. While the current Webadmin layout can be improved further, I feel that using a Javascript framework for something as simple as this is overkill.
